### PR TITLE
Excel time to numeric

### LINKED ISSUE #245: Feature Request: excel_numeric_to_time
We currently have `excel_numeric_to_date()` to assist with converting dates and datetimes to POSIXct objects.  Can we create a similar function for `excel_numeric_to_time()` or perhaps more accurately `excel_time_to_numeric()`?

Excel stores times in a format as a fraction of a day, and an issue that I commonly have when opening an Excel file with times is that they are not consistently read correctly (this is also related to tidyverse/readxl#504).

In files that I read, the person providing the file often mixes multiple formats accidentally, so that when read in, I get a mixture like the following:

```r
c("9:04AM", "14:35", "0.395833333333", "1899-12-31 08:05")
```

The issue that I have is usually related to the mixture of formats due to imprecise data entry.  I think it would be a good pairing with the `excel_numeric_to_date()` function to have a function to help handle times.  The interface I was thinking of is below.  The input would be any of the formats above:

* a fraction between 0 (inclusive) and 1 (exclusive) that may be a character string.
* Time with AM/PM (must have AM/PM case insensitively at the end)
* 24-hour time
* A POSIXlt or POSIXct object on the day "1899-12-31"
    * Or a character string that looks like a POSIXlt/POSIXct object from this day, notably the string starts with "1899-12-31 " and then has a 24-hour time.

The return value would be a the number of seconds in the time (a value >=0 and <86400).  Another option for output could be a lubridate duration object (but that would require lubridate as a dependency).

```r
excel_time_to_numeric(time_value, round_seconds=TRUE)
```

If interested, I can make a PR.

### PR DESCRIPTION
**Pull request by billdenney**
Replaces #246 (by switching to the `main` branch) and fixes #245

### CODE REVIEW

**Comment by billdenney** [comment]
@sfirke , Hopefully this can make the next CRAN release.

**Comment by sfirke** [comment]
I can review this this week.

**Comment by billdenney** [comment]
@sfirke, I found myself needing this function today.  So: ping!  😄

**Comment by sfirke** [comment]
Ack sorry I have been so slow!  Thanks for the ping, I will try to review soon.

File: R/excel_time_to_numeric.R [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+      # SI numbers have to have the form [number]E-[number] becasue the number
+      # has to be between 0 and 1 and can't be bigger than 1.
+      si_number = "^[1-9](\\.[0-9]*)?E-[0-9]+$",
+      "12hr" = "^([0]?[1-9]|1[0-2]):([0-5][0-9])(?::([0-5][0-9]))? ?([AP]M)$",
+      "24hr" = "^([0-1]?[0-9]|2[0-3]):([0-5][0-9])(?::([0-5][0-9]))?$",
+      # The ".*?" at the end of POSIX is to allow for a time zone, but it allows
+      #   for imperfect parsing if there were just a date and a space.
+      # The the entire time is optional to allow for midnight which shows as
+      #   just the date and time zone.
+      POSIX = "1899-12-31 (?:([0-1]?[0-9]|2[0-3]):([0-5][0-9])(?::([0-5][0-9]))?)?.*?$"
```
Comment (sfirke) [inline R/excel_time_to_numeric.R:92]: Will it still match midnight with the trailing space after the 1899-12-31?

File: R/excel_time_to_numeric.R [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+    list(
+      number = "^0(\\.[0-9]*)?$",
+      # SI numbers have to have the form [number]E-[number] becasue the number
+      # has to be between 0 and 1 and can't be bigger than 1.
+      si_number = "^[1-9](\\.[0-9]*)?E-[0-9]+$",
+      "12hr" = "^([0]?[1-9]|1[0-2]):([0-5][0-9])(?::([0-5][0-9]))? ?([AP]M)$",
+      "24hr" = "^([0-1]?[0-9]|2[0-3]):([0-5][0-9])(?::([0-5][0-9]))?$",
+      # The ".*?" at the end of POSIX is to allow for a time zone, but it allows
+      #   for imperfect parsing if there were just a date and a space.
+      # The the entire time is optional to allow for midnight which shows as
```
Comment (sfirke) [inline R/excel_time_to_numeric.R:90]:
```suggestion
      # The entire time is optional to allow for midnight which shows as
```

File: R/excel_time_to_numeric.R [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    ret[mask_clock] <-
+      as.numeric(hours[mask_clock]) * 3600 +
+      as.numeric(minutes[mask_clock]) * 60 +
+      as.numeric(seconds[mask_clock])
+  }
+  if (round_seconds) {
+    ret <- round(ret)
+  }
+  ret
+}
```
Comment (sfirke) [inline R/excel_time_to_numeric.R:166]: Elegant! 👏

File: tests/testthat/test-excel_time_to_numeric.R [hunk 4]
```diff
@@ -0,0 +1,162 @@
+context("Excel time cleaning")
```
Comment (sfirke) [inline tests/testthat/test-excel_time_to_numeric.R:1]: I still need to try out the tests and review them but don't want that to hold up my feedback on the actual code.

**Review by sfirke — COMMENTED** [review]
I looked at the actual code, not the tests yet.  Code looks great, all that thinking about time format regex has me 😵‍💫

File: R/excel_time_to_numeric.R (hunk 3 above)
Comment (sfirke) [reply R/excel_time_to_numeric.R:166]: The function overall, that is.

File: R/excel_time_to_numeric.R [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+#'
+#' \code{time_value} may be one of the following formats:
+#' \itemize{
+#'   \item{numeric}{The input must be a value from 0 to 1 (exclusive of 1); this value is returned as-is.}
+#'   \item{POSIXlt or POSIXct}{The input must be on the day 1899-12-31 (any other day is an error).  The time of day is extracted and converted to a fraction of a day.}
+#'   \item{character}{Any of the following (or a mixture of the choices):}
+#'   \itemize{
+#'     \item{A character string that is a number between 0 and 1 (exclusive of 1), converted like a numeric value.}
+#'     \item{A character string that looks like a date on 1899-12-31 (specifically, it must start with \code{"1899-12-31 "}), converted like a POSIXct object.}
+#'     \item{A character string that looks like a time.  Choices are 12-hour time ""}
```
[the remark this answers is not in the document]
Comment (billdenney) [reply R/excel_time_to_numeric.R:14]: Yes, I fixed that.

File: R/excel_time_to_numeric.R [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    rep(NA_real_, length(time_value))
+  } else {
+    stop("If given as a logical vector, all values must be ")
+  }
+}
+
+#' @export
+excel_time_to_numeric.numeric <- function(time_value, round_seconds = TRUE) {
+  if (all(is.na(time_value) |
+    time_value >= 0 &
```
[the remark this answers is not in the document]
Comment (billdenney) [reply R/excel_time_to_numeric.R:41]: Good point

File: R/excel_time_to_numeric.R (hunk 1 above)
Comment (billdenney) [reply R/excel_time_to_numeric.R:92]: I added this test which I think confirms that the answer is yes:

```
expect_equal(excel_time_to_numeric(as.POSIXct("1899-12-31 00:01")), 60)
```

**Comment by billdenney** [comment]
I think that I've addressed all the comments.  Can you please take one more look and merge if it looks good to you?

File: R/excel_time_to_numeric.R (hunk 3 above)
Comment (billdenney) [reply R/excel_time_to_numeric.R:166]: Thanks!

File: tests/testthat/test-excel_time_to_numeric.R [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+  expect_equal(excel_time_to_numeric("1899-12-31 08:00 foo"), 8 * 3600)
+  expect_equal(excel_time_to_numeric("1899-12-31 08:10 foo"), 8 * 3600 + 10 * 60)
+  expect_equal(excel_time_to_numeric("1899-12-31 8:10:05 foo"), 8 * 3600 + 10 * 60 + 5)
+  expect_equal(excel_time_to_numeric("1899-12-31 21:05 foo"), 21 * 3600 + 5 * 60)
+  expect_equal(excel_time_to_numeric("1899-12-31 0:05 foo"), 5 * 60)
+  expect_equal(excel_time_to_numeric("1899-12-31 00:05 foo"), 5 * 60)
+  expect_equal(excel_time_to_numeric("1899-12-31 21:05:20 foo"), 21 * 3600 + 5 * 60 + 20)
+})
+
+test_that("excel_time_to_numeric, POSIX times treat no time as midnight but only if there is a space indicating a mostly-well-formed date-time object.", {
```
Comment (sfirke) [inline tests/testthat/test-excel_time_to_numeric.R:142]: I feel like there must be some story behind this use case 😆

**Review by sfirke — APPROVED** [review]

**Comment by sfirke** [comment]
@billdenney this looks great!  Thank you very much, both for the great code and the patience as I was slow to review.

Feel free to squash and merge.  I thought one block of tests was redundant and wanted to leave the chance for you to remove it (if I'm not missing something) before merging, but this is ready to go.

I wonder if it's a good impetus to start the next CRAN submission, to make this available to a broader user base.

File: tests/testthat/test-excel_time_to_numeric.R (hunk 7 above)
Comment (billdenney) [reply tests/testthat/test-excel_time_to_numeric.R:142]: You'd be amazed at the trash that I have to sort through in some of my source data.  (Or maybe as the originator of `janitor`, you wouldn't 😄 .)  I sometimes get fields that have multiple data points concatenated together.  Dates and times are especially bad.  As long as it fully matches an expected format (maybe with extra bits at the end), I wanted to let it in.

**Pull request merged**